src/util/gmp_util.h

```diff
@@ -33,6 +41,49 @@ inline size_t gmpz_hash(const mpz_t toHash) {
   return hash;
 }/* gmpz_hash() */
 
+#ifdef CVC5_NEED_INT64_T_OVERLOADS
+/*
+ * Before converting an int64_t/uint64_t to a GMP value, we need to check if
+ * the input value will fit inside of a (un)signed long, which is the largest
+ * overload supported by GMP's C++ API.
+ *
+ * If the input value is *larger* than (un)signed long, then we use GMP's
+ * string-based API to construct the value.
+ *
+ * The following function looks-up the largest matching type against the input
+ * type, checks the range, and then decides to use the string-based overloads
+ * if the value does not fit.
+ */
+template <typename T>
+mpz_class construct_mpz(T z)
+{
+  /* only use this with int64_t or uint64_t */
+  static_assert(
+      std::is_same<T, int64_t>::value || std::is_same<T, uint64_t>::value,
+      "construct_mpz only supports int64_t or uint64_t");
+
+  /* what's the largest GMP-supported type? */
+  using gmp_type = typename std::
+      conditional<std::is_signed<T>::value, signed long, unsigned long>::type;
+
+  /* result value */
+  mpz_class result;
+
+  /* if our input fits inside of the range of the largest GMP-support type, we
+   * construct result based on the value */
+  if (std::numeric_limits<gmp_type>::min() <= z
+      && z <= std::numeric_limits<gmp_type>::max())
+  {
+    result = static_cast<gmp_type>(z);
+  }
+  else
+  {
+    /* otherwise, use the string-based methods */
+    result = std::to_string(z);
```

Comment:
I'm actually not sure how expensive the string conversion is, but since we are converting from [u]int64_t, we could also do the following, similar to what we do in Bitwuzla (https://github.com/bitwuzla/bitwuzla/blob/main/src/lib/bv/bitvector.cpp#L120):

for unsigned:
```
    uint32_t hi = static_cast<uint32_t>(z >> 32);
    uint32_t lo = static_cast<uint32_t>(z);
    mpz_init_set_ui(result, hi);
    mpz_mul_2exp(result, result, 32);
    mpz_add_ui(result, result, lo);
```

for signed:
```
    int32_t hi = static_cast<int32_t>(z >> 32);
    int32_t lo = static_cast<int32_t>(z);
    mpz_init_set_si(result, hi);
    mpz_mul_2exp(result, result, 32);
    mpz_add_ui(result, result, lo);
```

Reply:
So, yes, that’s what I originally thought I’d be doing but using mpz_import (https://github.com/cvc5/cvc5/pull/10334#issuecomment-1936727239), but then I saw that cvc5 was already using the string-based method.

As a note, I didn’t write the string-based code, it was already in cvc5, I just moved it to make it usable across multiple classes.

Do you want me to move this code to be using mpz_import or similar, or leave it as-is, given cvc5 already uses the string-based approach in main?

Reply:
Ok if it was already there, let's leave it as it is now.

Reply:
Yep: https://github.com/cvc5/cvc5/blob/2576488d68ea43470f152ab4c6d7785b9f9fb51e/src/util/integer_gmp_imp.cpp#L52 (main at the time of writing).

Basically, everywhere that cvc5 goes from an `(u)int64` -> GMP-type needs to go through that check (or do something like you're suggesting).

I went for the simplest route of just making that functionality common, and then using it everywhere that GMP types are instantiated from `(u)int64`.